feat: conditional formatting color range

IrakliJani opened the pull request:
Closes: #3936 

### Description:

acceptance criteria:
- [x] when I create a rule, I can pick if it's a `single color` rule or a `color range` rule
- [x] For a range of colours rule, you need to apply the following settings:
    - [x] Select a colour that you want to have for the minimum values in your colour range
    - [x] Select a colour that you want to have for the maximum values in your colour range.
    - [x] Select the minimum value that you want the range to start at.
    - [x] Select the maximum value for the range. 
- [x] Ranges are set to 5 colours (separate issue for setting a custom number of values in the colour range here: https://github.com/lightdash/lightdash/issues/3764).
- [x] for now, you can only set the min and max colour.

persisted configuration:

```js
conditionalFormattings: [
  // for single color range and conditional operator
  {
    "color": "#fac858",
    "rules": [
      {
        "id": "90f8b88a-b623-4853-9a9b-9ba460e7de4b",
        "values": [
          30
        ],
        "operator": "greaterThan"
      }
    ],
    "target": {
      "fieldId": "orders_total_order_amount"
    }
  },
  // for multi color range and min/max values
  {
    "rule": {
      "max": 40,
      "min": 30
    },
    "color": {
      "end": "#73c0de",
      "start": "#b39e6f",
      "steps": 5 // <--------- hardcoding `5` for now (generates 5 colors between `start` and `end`
    },
    "target": {
      "fieldId": "orders_customer_id"
    }
  }
]

```

video demo:

https://github.com/lightdash/lightdash/assets/962095/e83a4502-1e1e-469f-9711-bc412ea4a55b

in results table:
<img width="874" alt="CleanShot 2023-07-17 at 16 17 00@2x" src="https://github.com/lightdash/lightdash/assets/962095/62d38599-b5c5-4e07-96b9-498cdd7c96b3">

in pivot tables:
<img width="1157" alt="CleanShot 2023-07-17 at 16 13 48@2x" src="https://github.com/lightdash/lightdash/assets/962095/e7433bba-6981-4229-8f42-014da6dbb454">

rephus commented on packages/frontend/src/components/VisualizationConfigs/TableConfigPanel/ConditionalFormatting.tsx:
```diff
@@ -49,82 +66,181 @@ const ConditionalFormatting: FC<ConditionalFormattingProps> = ({
+            if (isConditionalFormattingConfigWithSingleColor(config)) {
+                handleChange(
+                    produce(config, (draft) => {
+                        // FIXME: check if we can fix this problem in number input
```
do we still need to fix this ?

IrakliJani replied:
yep, unfortunately this comes from blueprint input which is in filter components (not part of this change)

rephus reviewed (commented):
Code looks good, let me test this on render

TuringLovesDeathMetal commented:
weooooo!!! Looks awesome 😍 Super excited for this!

The functionality is great! It works as expected which is awesome :) Just a few things I noticed (but if these require a lot of work, let me know and I'll rethink if we actually need them):

- Can we add the separator text/subheading `select type` (this should look the same as `select field` and `select colour`) above the `single colour` vs. `colour range`? 
<img width="324" alt="image" src="https://github.com/lightdash/lightdash/assets/31848148/9ade48ab-0f56-4297-9b80-47f653ada71e">

- Can we have colours applied by default to the range? So, white for min value, then the first colour in their default palette for the max value? 
<img width="335" alt="image" src="https://github.com/lightdash/lightdash/assets/31848148/2a71f160-b003-4001-a839-c26d75e38c52">

- Can we have the max value set by default to the max value in the column they're applying the conditional formatting to? (if this would take more than 30 mins, then let's not do it) 
<img width="167" alt="image" src="https://github.com/lightdash/lightdash/assets/31848148/148177bd-66df-4f2c-b413-2b79c9aeb3df">
 
- okay, this one might be complicated...but on hover, you see the rule, and it says `greater than (min value) and less than (max value)`. I'd expect it to say what colour it is based on the "rules"/intervals we've set in the colour range. E.g. if white = values between [0, 5], I'd expect to see that when hovering over white. Not `values between [0, 100]`.
<img width="429" alt="image" src="https://github.com/lightdash/lightdash/assets/31848148/e01c0f9e-bc71-4dc3-886d-daffb3acb3fe">

- I don't think this is something that was introduced with the ranges...but one thing I found weird was how my conditional formatting disappeared if I made my table into a pivot table 🤔 I'm not moving the column that has the conditional formatting rule...so it feels weird that the conditional formatting disappears once I've added a pivot. https://www.loom.com/share/f8d539c0044a4c11b4b58bc7ef844b8f

IrakliJani commented:
@TuringLovesDeathMetal thanks for a review!
- [x] adjusted defaults
- [x] added label "Select type" to a input

re min/max: it's not going to be a simple addition so I'm gonna spec a new ticket for it!

IrakliJani commented:
@TuringLovesDeathMetal 

re pivot tables: we don't apply conditional formatting to columns that are either in header or indexes. only for values. this is a "bug" in a sense that we allow conditional formatting to be added to that columns, or this can be a separete "feature request" to support cond-formatting to header/indexes in pivot tables.

p.s. specced ticket for min/max auto setting https://github.com/lightdash/lightdash/issues/6406

TuringLovesDeathMetal reviewed (approved):
Looks great!
